### PR TITLE
Prepare documentation for v0.1 release

### LINKED ISSUE #194: Document the “why”
**Is your feature request related to a problem? Please describe.**
Currently, Topiary doesn't advertise why it exists, what problem it solves. Of which there are two parts, probably: why should I want to use Topiary to format my code (which may have a rather boring answer), and why should I use Topiary to create a new formatter?

**Describe the solution you'd like**
Currently the README is the entirety of the documentation. We should add a new section to the README, right after the introductory paragraph and before the _Getting Started_ section. This section should answer the above two questions. In, say, one or two sentences each (it can be more, but doesn't need to, I don't think).

### LINKED ISSUE #195: Document the design principles
**Is your feature request related to a problem? Please describe.**

The README does a good job of documenting the technical design of Topiary. What the engine does, the command it takes. And even how to use Topiary productively.

What is absent as far as I can tell is the design _principles_ under which we've been operating. Things like “try to minimise diffs” (which is a principle for individual languages, for the most part, but we also need to provide the tools for that), the multiline vs single line.

**Describe the solution you'd like**

Have a _Design principles_ section in the README. Probably before the _Getting Started_ section (but after the whys of #194 ). In this section we list the constraint that we put on ourselves, and are guiding the design of the tool. We can reuse some of [those from Ormolu](https://github.com/tweag/ormolu/#readme).

### PR DESCRIPTION
**Pull request by Xophmeister**
This PR:

* [x] Tidies up the original `README`.
* [x] Resolves #194.
* [x] Resolves #195.
* [x] Documents the current state of language support.

**I will defer merging until this has at least two approvals :heavy_check_mark:**

### CODE REVIEW

**Review by nbacquey — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,11 +1,87 @@
...
+
+## Motivation
+
+The style in which code is written has, historically, been mostly left
+to personal choice. Of course, this is subjective by definition and has
+led to many wasted hours reviewing formatting choices, rather than the
+code itself. Prescribed style guides were an early solution to this,
+spawning tools that lint a developer's formatting and ultimately leading
+to automatic formatters. The latter were popularised by
+[`gofmt`][gofmt], whose developers had the insight that "good enough"
```
Comment (aspiwack) [inline README.md:17]: Maybe we ought to link to the slides which talk about design decisions somewhere.

For the record, I don't care too much about the absence of configurability (in fact Topiary is infinitely configurable: just clone the query file, and do the modifications that you want). It is important in Ormolu only because Haskell is so hard to parse and format that it's quite hard to test with configuration options (yes, changing from 2 to 4 spaces of indentation can conceivably break a Haskell output if not done carefully) but here, it doesn't matter too much.

File: README.md [hunk 2]
```diff
@@ -1,11 +1,87 @@
...
+![CI Status][ci-badge]
 
 Topiary aims to be a uniform formatter for simple languages, as part of
-the [Tree-sitter](https://tree-sitter.github.io/tree-sitter/) ecosystem.
-It is named after the art of clipping or trimming trees into fantastic
-shapes.
+the [Tree-sitter] ecosystem. It is named after the art of clipping or
+trimming trees into fantastic shapes.
+
+## Motivation
```
Comment (aspiwack) [inline README.md:9]: Before even this section we need a section explaining to the reader in which case he should care about Topiary. Something brief and to the point. There are two typical users: formatter authors, and formatter users. The value proposition is easier to articulate for authors, evidently. For users, we can explain the formatting philosophy (which we describe in more details below), and maybe the fact that it can conveniently serve as a formatter for many languages in their repository (ore repositories even).

File: README.md [hunk 3]
```diff
@@ -1,11 +1,87 @@
...
+rapid prototyping of formatters, providing a [Tree-sitter
+grammar][tree-sitter-parsers] is defined for that language.
+
+## Design Principles
+
+Topiary has been created with the following goals in mind:
+
+* To target languages for which a [Tree-sitter
+  grammar][tree-sitter-parsers] exist; whether officially supported by
+  the Tree-sitter project, or otherwise.
```
Comment (aspiwack) [inline README.md:34]: This paragraph sounds a bit cheesy. I guess we can say “use tree-sitter for parsing” if we want to make this point. Maybe explain that this avoids making yet another grammar for a formatter.

File: README.md [hunk 4]
```diff
@@ -1,11 +1,87 @@
...
+
+* To support a handful of languages, with "one true formatting style"
+  for each, where no de facto formatter exists. These supported
+  formatting styles must:
```
Comment (aspiwack) [inline README.md:38]: As said above, “one true formatting style” is not a target. At any rate, they never exist in languages that have lived a while before their formatter was out. At any rate², “support[ing] a handful of languages” is not a design principle. It was a target for our MVP.

File: README.md [hunk 5]
```diff
@@ -1,11 +1,87 @@
...
+
+  * Be compatible with attested formatting styles used for that language
+    in the wild.
+
+  * Be faithful to the author's intent, in terms of line spacing.
```
Comment (aspiwack) [inline README.md:43]: I don't know what this means.

File: README.md [hunk 6]
```diff
@@ -1,11 +1,87 @@
...
+
+  * Result in minimal diffs.
```
Comment (aspiwack) [inline README.md:45]: I know what this means. But I doubt a typical reader will.

File: README.md [hunk 7]
```diff
@@ -1,11 +1,87 @@
...
+
+  * Be idempotent. That is, formatting of already-formatted code doesn't
+    change anything.
+
+* Code must be written in such a way that it's easy to modify and
+  maintain.
+
```
Comment (aspiwack) [inline README.md:52]: Don't state the obvious, I feel it's insulting the readers' intelligence.

```suggestion
```

File: README.md [hunk 8]
```diff
@@ -1,11 +1,87 @@
...
+* Code and formatting styles must be well-tested and robust, so that the
+  formatter can be used in large projects.
+
+* For end users -- i.e., not formatting style developers -- the
+  formatter should:
+
+  * Not permit any subjective configuration.
```
Comment (aspiwack) [inline README.md:59]: Again :slightly_smiling_face:

File: README.md [hunk 9]
```diff
@@ -1,11 +1,87 @@
...
+
+#### Supported
+* OCaml (both implementations and interfaces)
+* JSON
+* TOML
+
+#### Experimental
+
+* Rust
+* Bash
```
Comment (aspiwack) [inline README.md:83]: Be fair to your languages and give all of them a link.

File: README.md [hunk 10]
```diff
@@ -1,11 +1,87 @@
...
+* [Nickel][nickel]
```
Comment (aspiwack) [inline README.md:84]: You will be amused to know that 
```suggestion
* [Nickel]
```
works, because link names are case-insensitive in Markdown (I discovered recently).

File: README.md [hunk 11]
```diff
@@ -35,31 +111,32 @@ a development environment.
 Options:
 
 * `-l`, `--language <LANGUAGE>`\
-  Which language to parse and format [possible values: json, toml]
+  Which language to parse and format [possible values: `json`, `toml`,
+  `ocaml`, `ocaml-implementation`, `ocaml-interface`].
```
Comment (aspiwack) [inline README.md:115]: Is this part of the README generated?

File: README.md [hunk 12]
```diff
@@ -624,69 +732,96 @@ This Tree-sitter query:
   (#scope_id! "tuple")
 )
 ```
-formats this piece of code:
+
+...formats this piece of code:
```
Comment (aspiwack) [inline README.md:736]: I don't care either way, but this looks like a pretty unusual typographic style.

**Review by aspiwack — COMMENTED** [review]
Comments below.

As a side note: there are a bunch of rewrapped paragraph in this PR. It is preferable to do paragraph rewrapping in separate PRs (just like any formatting of code) so as to facilitate review.

File: README.md (hunk 5 above)
Comment (Xophmeister) [reply README.md:43]: It's a paraphrasing of a point made in the `gofmt` slides:

> Respecting user intent is key.

i.e., If the author has written code that spans multiple lines, then they must have done so for a reason and that ought to be preserved. Hence Topiary's single-line vs. multi-line distinction.

File: README.md (hunk 6 above)
Comment (Xophmeister) [reply README.md:45]: (This was copied from Ormolu's `README`, but I will expand...)

File: README.md (hunk 7 above)
Comment (Xophmeister) [reply README.md:52]: (This was copied from Ormolu's `README`, but I will remove it.)

File: README.md (hunk 11 above)
Comment (Xophmeister) [reply README.md:115]: No. It probably should be... I'll add it as an issue.

File: README.md (hunk 12 above)
Comment (Xophmeister) [reply README.md:736]: I don't like the look of what appears to be a new sentence, but is actually just a follow-on clause interjected with a code block, starting with a lowercase letter. I add ellipses in this case as an indicator of this.

File: README.md (hunk 11 above)
Comment (Xophmeister) [reply README.md:115]: Done: Issue #277

File: README.md (hunk 8 above)
Comment (Xophmeister) [reply README.md:59]: IMO, the lack of configurability -- as an end user -- is important; its necessary for delivering on uniformity. Obviously the styles are customisable if you're so inclined, but I think that doing so pushes you into a different user category.

File: README.md (hunk 1 above)
Comment (Xophmeister) [reply README.md]: Added in fd18ebc1c1a

File: README.md (hunk 2 above)
Comment (Xophmeister) [reply README.md:15]: Added in fd18ebc1c1a

File: README.md (hunk 3 above)
Comment (Xophmeister) [reply README.md]: Changed in fd18ebc1c1a

File: README.md (hunk 4 above)
Comment (Xophmeister) [reply README.md]: Changed in fd18ebc1c1a

File: README.md (hunk 5 above)
Comment (Xophmeister) [reply README.md]: Reworded in fd18ebc1c1a

File: README.md (hunk 6 above)
Comment (Xophmeister) [reply README.md]: Reworded in fd18ebc1c1a

File: README.md (hunk 7 above)
Comment (Xophmeister) [reply README.md]: Removed in fd18ebc1c1a

File: README.md (hunk 8 above)
Comment (Xophmeister) [reply README.md]: Reworded in fd18ebc1c1a

File: README.md (hunk 9 above)
Comment (Xophmeister) [reply README.md]: Added in fd18ebc1c1a

File: README.md (hunk 10 above)
Comment (Xophmeister) [reply README.md]: Done in fd18ebc1c1a

File: README.md [hunk 13]
```diff
@@ -1,11 +1,99 @@
...
 Topiary aims to be a uniform formatter for simple languages, as part of
-the [Tree-sitter](https://tree-sitter.github.io/tree-sitter/) ecosystem.
-It is named after the art of clipping or trimming trees into fantastic
-shapes.
+the [Tree-sitter] ecosystem. It is named after the art of clipping or
+trimming trees into fantastic shapes.
+
+Topiary is designed for formatter authors and formatter users. Authors
+can create a formatter for a language without having to write their own
+parser. Users benefit from uniform code style and, potentially, the
```
Comment (aspiwack) [inline README.md:11]: Perhaps more importantly than the parser, is the whole engine as well.

I'm not sure the case for users is particularly good, but then again, we have little to say about that, I suppose. Maybe something along the line of “a single formatter for many languages with comparable style between languages”. After all, in our related work, we have a cornucopia of multi-language formatters. So it seems that's a value.

File: README.md [hunk 14]
```diff
@@ -1,11 +1,99 @@
...
+
+The style in which code is written has, historically, been mostly left
+to personal choice. Of course, this is subjective by definition and has
+led to many wasted hours reviewing formatting choices, rather than the
+code itself. Prescribed style guides were an early solution to this,
+spawning tools that lint a developer's formatting and ultimately leading
+to automatic formatters. The latter were popularised by
+[`gofmt`][gofmt], whose developers had [the insight][gofmt-slides] that
+"good enough" uniform formatting, imposed on a codebase, with no scope
+for subjective adjustment, largely resolves these problems.
```
Comment (aspiwack) [inline README.md:25]:
```suggestion
[`gofmt`][gofmt], whose developers had [the insight][gofmt-slides] that
"good enough" uniform formatting, imposed on a codebase largely resolves these problems.
```
I don't think that this bits adds relevant information.

File: README.md [hunk 15]
```diff
@@ -1,11 +1,99 @@
...
+
+Topiary follows this trend by aspiring to be a "universal formatter
+engine", which allows developers to not only automatically format their
+codebases with a uniform style, but to define that style for new
+languages using a [simple DSL][tree-sitter-query]. This allows for the
+rapid prototyping of formatters, providing a [Tree-sitter
```
Comment (aspiwack) [inline README.md:31]:
```suggestion
fast development of formatters, providing a [Tree-sitter
```

File: README.md [hunk 16]
```diff
@@ -1,11 +1,99 @@
...
+
+## Design Principles
+
+Topiary has been created with the following goals in mind:
+
+* Use [Tree-sitter] for parsing, to avoid writing yet another grammar
+  for a formatter.
+
+* For bundled formatting styles -- for a handful of languages where no
+  de facto formatter exists -- to meet the following constraints:
```
Comment (aspiwack) [inline README.md:42]:
```suggestion
* For bundled formatting styles to meet the following constraints:
```

File: README.md [hunk 17]
```diff
@@ -1,11 +1,99 @@
...
+
+  * Be compatible with attested formatting styles used for that language
+    in the wild.
+
+  * Be faithful to the author's intent: If code has been written such
```
Comment (aspiwack) [inline README.md:47]:
```suggestion
  * Be faithful to the author's intent: if code has been written such
```

File: README.md [hunk 18]
```diff
@@ -1,11 +1,99 @@
...
+    that it spans multiple lines, that decision is preserved.
+
+  * Minimise changes between commits such that diffs focus mainly on the
+    code that's changed, rather that superficial artefacts.
```
Comment (aspiwack) [inline README.md:51]: I'm not sure that this is a good description of diff minimisation. The diff minimisation goal is to make a style where 1/ change on one line doesn't format others (I think that's what you are describing) 2/ (and, I'd say, it is more often where arbitrage is) a formatting style that doesn't require you to modify a line spuriously.

(2) is the reason for normalising terminators when we can (we currently can't in Ocaml, but this is why this style is desired):
```ocaml
[
  1;
  2;
]
```

File: README.md [hunk 19]
```diff
@@ -708,3 +855,57 @@ to find the bundled integration test input file for the given language.
...
+* [null-ls.nvim]: An LSP framework for Neovim that facilitates formatter
+  orchestration.
+* [prettier]: A formatter with support for multiple (web-development
+  related) languages.
+* [treefmt]: A general formatter orchestrator, which unifies formatters
+  under a common interface.
+
+### Related Formatters
+
+* [gofmt]: A (_the_) formatter for Go.
```
Comment (aspiwack) [inline README.md:878]: And major inspiration for the style of our formatters.

File: README.md [hunk 20]
```diff
@@ -708,3 +855,57 @@ to find the bundled integration test input file for the given language.
...
+* [ocamlformat]: A formatter for OCaml.
+* [ocp-indent]: A tool to indent OCaml code.
+* [Ormolu]: A formatter for Haskell that follows similar design
```
Comment (aspiwack) [inline README.md:881]:
```suggestion
* [Ormolu]: Our formatter for Haskell that follows similar design
```

File: README.md [hunk 21]
```diff
@@ -708,3 +855,57 @@ to find the bundled integration test input file for the given language.
...
+  principles as Topiary.
+* [rustfmt]: A formatter for Rust.
```
Comment (aspiwack) [inline README.md:883]:
```suggestion
* [rustfmt]: The _de facto_ standard formatter for Rust.
```

File: README.md (hunk 13 above)
Comment (Xophmeister) [reply README.md]: Added in cd20b49a278a

File: README.md (hunk 14 above)
Comment (Xophmeister) [reply README.md]: Removed in cd20b49a278a

File: README.md (hunk 15 above)
Comment (Xophmeister) [reply README.md]: Changed in cd20b49a278a

File: README.md (hunk 16 above)
Comment (Xophmeister) [reply README.md]: Removed in cd20b49a278a

File: README.md (hunk 17 above)
Comment (Xophmeister) [reply README.md]: Changed in cd20b49a278a

File: README.md (hunk 18 above)
Comment (Xophmeister) [reply README.md]: Expanded in cd20b49a278a

File: README.md (hunk 19 above)
Comment (Xophmeister) [reply README.md]: Added in cd20b49a278a

File: README.md (hunk 20 above)
Comment (Xophmeister) [reply README.md]: Changed in cd20b49a278a

File: README.md (hunk 21 above)
Comment (Xophmeister) [reply README.md]: Added in cd20b49a278a

**Comment by Xophmeister** [comment]
Thanks, @aspiwack :pray: I've added those finishing touches in cd20b49a278a

**Review by aspiwack — APPROVED** [review]
